test/source-sink-errors/mzcompose.py

```diff
@@ -0,0 +1,172 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+import random
+from dataclasses import dataclass
+from textwrap import dedent
+from typing import Callable, Optional
+
+from materialize.mzcompose import Composition
+from materialize.mzcompose.services import Materialized, Redpanda, Storaged, Testdrive
+
+SERVICES = [Redpanda(), Materialized(), Testdrive(), Storaged()]
+
+
+@dataclass
+class Disruption:
+    name: str
+    breakage: Callable
+    expected_error: str
+    fixage: Optional[Callable]
+
+    def run_test(self, c: Composition) -> None:
+        print(f"+++ Running disruption scenario {self.name}")
+        seed = random.randint(0, 256**4)
+
+        c.down(destroy_volumes=True)
+        c.up("testdrive", persistent=True)
+        c.start_and_wait_for_tcp(services=["redpanda", "materialized", "storaged"])
+        c.wait_for_materialized()
+
+        with c.override(
+            Testdrive(
+                no_reset=True,
+                seed=seed,
+                entrypoint_extra=["--initial-backoff=1s", "--backoff-factor=0"],
+            )
+        ):
+            self.populate(c)
+            self.breakage(c, seed)
+            self.assert_error(c, self.expected_error)
+
+            if self.fixage:
+                self.fixage(c, seed)
+                self.assert_recovery(c)
+
+    def populate(self, c: Composition) -> None:
+        # Create a source and a sink
+        c.testdrive(
+            dedent(
+                """
+                > CREATE CONNECTION kafka_conn
+                  TO KAFKA (BROKER '${testdrive.kafka-addr}');
+
+                > CREATE CONNECTION IF NOT EXISTS csr_conn TO CONFLUENT SCHEMA REGISTRY (
+                  URL '${testdrive.schema-registry-url}'
+                  );
+
+                $ kafka-create-topic topic=source-topic
+
+                $ kafka-ingest topic=source-topic format=bytes
+                ABC
+
+                > CREATE SOURCE source1
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-source-topic-${testdrive.seed}')
+                  FORMAT BYTES
+                  ENVELOPE NONE
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
+
+                > CREATE SINK sink1 FROM source1
+                  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink-topic-${testdrive.seed}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE DEBEZIUM
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
+
+                # Wait for the source to start up
```

Comment:
@bkirwi If the disruption happens immediately after `CREATE SOURCE` succeeds, without waiting for the first data to arrive, the status table never transitions away from the `started` state, even though the logs contains the expected `librdkafka` errors.

It is true that in some sense the source never properly got "running", but being stuck in the "started" state and not in "stalled" means that the error column is empty even though librdkafka is already reporting the error in the log.

Reply:
I agree that `stalled` would be better here... looking into it.